From d0321119e5ab585d509e3b74aa3138fb089d756c Mon Sep 17 00:00:00 2001
From: Wright Feng <wright.feng@cypress.com>
Date: Wed, 21 Nov 2018 20:42:39 -0600
Subject: [PATCH 035/101] brcmfmac: clean up iface mac descriptor before
 de-initializing it

We saw following warning message after removing USB Wi-Fi dongle with
data connection. The root cause is that some skb were queued in iface
mac descriptor psq and occupied the hanger slots, and we didn't de-queue
the skb before de-initializing iface mac descriptor. So it triggered
WARN_ON when cleaning up hanger slot eventually.

backports/drivers/net/wireless/broadcom/brcm80211/brcmutil/utils.c:49
brcmu_pkt_buf_free_skb+0x21/0x30 [brcmutil]
[  977.523200] Modules linked in: brcmfmac(OE) brcmutil(OE) cfg80211(OE)
compat(OE) sdhci_pci(OE) sdhci(OE) mmc_core(OE) rfkill ip6table_filter
ip6_tables ebtable_nat ebtables dns_resolver fscache e1000e mei_me mei
tpm_tis tpm_tis_core tpm iTCO_wdt ppdev iTCO_vendor_support pcspkr
lpc_ich i2c_i801 mfd_core ptp pps_core parport_pc parport wmi uinput
tcp_bic i915 iosf_mbi i2c_algo_bit drm_kms_helper drm i2c_core video
[last unloaded: brcmfmac]
[  977.523219] CPU: 1 PID: 1306 Comm: kworker/1:1 Tainted: G
OE   4.12.0 #1
[  977.523220] Hardware name:                  /DH77EB, BIOS
EBH7710H.86A.0100.2013.0312.1351 03/12/2013
[  977.523223] Workqueue: usb_hub_wq hub_event
[  977.523224] task: ffff880118703600 task.stack: ffffc90000be4000
[  977.523226] RIP: 0010:brcmu_pkt_buf_free_skb+0x21/0x30 [brcmutil]
[  977.523227] RSP: 0018:ffffc90000be7a98 EFLAGS: 00010086
[  977.523228] RAX: 0000000000000045 RBX: ffffffffa03fa850 RCX:
0000000000000006
[  977.523228] RDX: 0000000000000000 RSI: 0000000000000092 RDI:
ffff8801106a3ce8
[  977.523229] RBP: ffffc90000be7a98 R08: 0000000000000000 R09:
000000000000045b
[  977.523229] R10: 0000000000000001 R11: 0000000000aaaaaa R12:
0000000000000988
[  977.523230] R13: ffff880110569938 R14: 0000000000000002 R15:
ffff8801106a3ce8
[  977.523231] FS:  0000000000000000(0000) GS:ffff88011f280000(0000)
knlGS:0000000000000000
[  977.523232] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  977.523232] CR2: 00007f7873490000 CR3: 0000000001c09000 CR4:
00000000001406e0
[  977.523233] Call Trace:
[  977.523244]  brcmf_fws_cleanup+0x1f2/0x230 [brcmfmac]
[  977.523251]  brcmf_fws_del_interface+0x70/0x90 [brcmfmac]
[  977.523257]  brcmf_proto_bcdc_del_if+0xe/0x10 [brcmfmac]
[  977.523262]  brcmf_remove_interface+0x69/0x190 [brcmfmac]
[  977.523267]  brcmf_detach+0x90/0xe0 [brcmfmac]
[  977.523273]  brcmf_usb_disconnect+0x76/0xb0 [brcmfmac]
[  977.523275]  usb_unbind_interface+0x72/0x260
[  977.523279]  device_release_driver_internal+0x141/0x200
[  977.523280]  device_release_driver+0x12/0x20
[  977.523282]  bus_remove_device+0xfd/0x170
[  977.523283]  device_del+0x1d9/0x300
[  977.523284]  usb_disable_device+0x9e/0x270
[  977.523286]  usb_disconnect+0x94/0x270
[  977.523287]  usb_disconnect+0x1f1/0x270
[  977.523288]  hub_event+0x76d/0x13b0
[  977.523291]  process_one_work+0x144/0x360
[  977.523293]  worker_thread+0x4d/0x3c0
[  977.523294]  kthread+0x112/0x150
[  977.523295]  ? rescuer_thread+0x340/0x340
[  977.523296]  ? kthread_park+0x60/0x60
[  977.523298]  ? SyS_exit_group+0x14/0x20
[  977.523300]  ret_from_fork+0x25/0x30
[  977.523301] Code: 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 48 85
ff 74 15 48 83 3f 00 55 48 89 e5 75 0d be 01 00 00 00 e8 82 5a 3e e1 5d
f3 c3 <0f> ff eb ef 66 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 55

Signed-off-by: Wright Feng <wright.feng@cypress.com>
Signed-off-by: Chi-hsien Lin <chi-hsien.lin@cypress.com>
---
 drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
index e81204d..03153be 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/fwsignal.c
@@ -2208,6 +2208,7 @@ void brcmf_fws_del_interface(struct brcmf_if *ifp)
 
 	brcmf_fws_lock(fws);
 	ifp->fws_desc = NULL;
+	brcmf_fws_macdesc_cleanup(fws, entry, ifp->ifidx);
 	brcmf_dbg(TRACE, "deleting %s\n", entry->name);
 	brcmf_fws_macdesc_cleanup(fws, &fws->desc.iface[ifp->ifidx],
 				  ifp->ifidx);
-- 
1.9.1

